sys/net/network_layer/ng_ipv6/ng_ipv6.c

```diff
@@ -0,0 +1,501 @@
+static void _ipv6_send(ng_pktsnip_t *pkt, bool prep_hdr)
+{
+    kernel_pid_t iface = KERNEL_PID_UNDEF;
+    ng_pktsnip_t *ipv6 = pkt, *payload;
+    ng_ipv6_hdr_t *hdr;
+    ng_ipv6_nc_t *nc_entry;
+
+    pkt = _start_write(pkt);
+
+    if (pkt == NULL) {
+        DEBUG("ipv6: unable to get write access to packet, dropping packet\n");
+        return;
+    }
+
+    /* get IPv6 snip and (if present) generic interface header */
+    if (pkt->type == NG_NETTYPE_NETIF) {
+        /* If there is already a netif header (routing protocols and
+         * neighbor discovery might add them to preset sending interface) */
+        iface = ((ng_netif_hdr_t *)pkt->data)->if_pid;
+        ipv6 = pkt->next;
+    }
+
+    hdr = ipv6->data;
+    payload = ipv6->next;       /* TODO: parse extension headers */
+
+    if (ng_ipv6_addr_is_multicast(&hdr->dst)) {
+        _ipv6_send_multicast(iface, pkt, ipv6, payload, prep_hdr);
+    }
+    else {
+        ng_ipv6_addr_t *next_hop = NULL;
+
+        next_hop = &hdr->dst;   /* TODO: next hop determination */
+
+        if ((nc_entry = ng_ipv6_nc_get_reachable(iface, next_hop)) == NULL) {
+            DEBUG("ipv6: No link layer address for next_hop %s found.\n",
+                  ng_ipv6_addr_to_str(addr_str, next_hop, sizeof(addr_str)));
+            ng_pktbuf_release(pkt);
+            return;
+        }
+        else {
+            iface = nc_entry->iface;
+        }
+
+        if (iface == KERNEL_PID_UNDEF) {
+            DEBUG("ipv6: no interface for %s registered, dropping packet\n",
+                  ng_ipv6_addr_to_str(addr_str, next_hop, sizeof(addr_str)));
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        if (_fill_ipv6_hdr(iface, ipv6, payload) < 0) {
+            /* error on filling up header */
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        _ipv6_send_unicast(iface, nc_entry->l2_addr, nc_entry->l2_addr_len, pkt);
+    }
+}
+
+/* functions for receiving */
+static inline bool _pkt_not_for_me(kernel_pid_t *iface, ng_ipv6_hdr_t *hdr)
+{
+    if (*iface == KERNEL_PID_UNDEF) {
+        *iface = ng_ipv6_netif_find_by_addr(NULL, &hdr->dst);
+        return (*iface == KERNEL_PID_UNDEF);
+    }
+    else {
+        return (ng_ipv6_netif_find_addr(*iface, &hdr->dst) == NULL);
+    }
+}
+
+static void _dispatch_ipv6_pkt(ng_nettype_t type, uint32_t demux_ctx,
+                               ng_pktsnip_t *pkt)
+{
+    msg_t msg;
+    ng_netreg_entry_t *entry = ng_netreg_lookup(type, demux_ctx);
+
+    msg.type = NG_NETAPI_MSG_TYPE_RCV;
+    msg.content.ptr = (char *)pkt;
+
+    while (entry) {
+        DEBUG("ipv6: Send receive command for %p to %" PRIu16 "\n", (void *)pkt,
+              entry->pid);
+        msg_send(&msg, entry->pid);
+        entry = ng_netreg_getnext(entry);
+    }
+}
+
+static void _ipv6_receive(kernel_pid_t iface, ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *ipv6, *payload;
```

Comment:
Because - at the point of commenting - I only saw it used as L2 payload, why a variable called `payload` in the IPv6 implementation file suggests that this refers to L2 payload. Since it is used for both, I think it is somehow fine.